src/Base/PyHandle.h

```diff
@@ -0,0 +1,152 @@
+ * implementation, don't use "new_reference".
+ *
+ * * when returning PyObject* from a getPyObject, use
+ * Py::new_reference_to(<PyHangle> object).
+ *
+ * * when returning PyObject* from a ...PyImp method implementation, use
+ * Py::new_reference_to(<PyHangle> object).
+ *
+ * The handle derives from PyCXX's Py::Object, so it can be conveniently used
+ * for adding to tuples and the like, when programming python API.
+ */
+ // Since it was made to become the main memory management construct of
+ // ConstraintSolver, checks for isValid() are intentionally not done to maximize
+ // performance. It is named "Unsafe" with the intention to have room for a safer
+ // "PyHandle" replica.
+template <class CppType>
+class UnsafePyHandle : public Py::Object
+{
+private:
+    CppType* cppptr;
+protected:
+    virtual void setObject(PyObject* pyob, bool new_reference = false){
+        if (typecheck(pyob))
+            cppptr = static_cast<CppType*>(static_cast<Base::PyObjectBase*>(pyob)->twinPtr());
+        else
+            throw Base::TypeError("Not a freecad object");
+        this->set(pyob, new_reference);
+    }
+
+public:
+    virtual bool typecheck(PyObject* pyob) {
+        if (! PyObject_TypeCheck(pyob, &(Base::PyObjectBase::Type)))
+            return false;
+        if (! static_cast<Base::PyObjectBase*>(pyob)->isValid())
+            return false;
+        return true;
+    }
+
+    //We can't override Py::Object::set
+    //so, mask out all methods to change pointer of Py::Object
+    //it may still be possible to bypass with typecasts,
+    //but at least that isn't straightforward.
+    /**
+     * @brief UnsafePyHandle constructor
+     * @param pyob
+     * @param new_reference: specify True if the reference is "owned" by the calling function,
+     * particularly when creating a new object like UnsafePyHandle<Base::Placement> hplm (new PlacementPy(...))
+     * For more info, see docs for PyCXX's Object's "owned" argument.
+     */
+    explicit UnsafePyHandle(PyObject* pyob, bool new_reference)
+    {
+        setObject(pyob, new_reference);
+    }
+
+    UnsafePyHandle( const Object &other )
+    {
+        setObject(other.ptr());
+    }
+
+    UnsafePyHandle& operator=(PyObject* pyob){
+        setObject(pyob);
+    }
+
+    UnsafePyHandle& operator=(Py::Object other){
+        setObject(other.ptr());
+    }
+
+    /**
+     * @brief isValid: if the C++ object was deleted but Py object is still
+     * around, isValid will return false. Attempting to dereference the handle
+     * in such state is dereferencing a null pointer (i.e., crash).
+     * @return if it's ok to dereference the handle.
+     */
+    bool isValid() const {
+        return static_cast<Base::PyObjectBase*>(this->ptr())->isValid();
+    }
+
+    CppType& operator*() const {
+        return *cppptr;
+    }
+    CppType* operator->() const {
+        return cppptr;
+    }
+
+    bool operator==(UnsafePyHandle<CppType> &other){return ptr() == other.ptr();}
```

Comment:
Due you truly wish to check that the two pointers are equivalent, i.e. are the same memory address? Or do you wish to check if the two objects being pointed to are equivalent?

Reply:
eeem... I'm not sure. Maybe it should be removed, then.

Reply:
I tend to leave code out until I need it - if your client code truly needs to check whether to `UnsafePyHandle` are equivalent, then you should definitely leave it. I'm not sure about the intended use, so I can't really comment on which way is "right".